test: Check if swupd is able to perform updates in last releases

otaviobp opened the pull request:
Tests added:
 - Check if bundle-list --all is working (because this affect the next tests)
 - Check if we can install an old version of clear and update interactively
   until last release
 - Check if we can install and old version and update to the most recent one

As this test takes a really long time to run, I'm not adding it to the
tests list to be executed to make check, but this can be used to validate
swupd before a release.

Signed-off-by: Otavio Pontes <otavio.pontes@intel.com>

phmccarty commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+	for v in $version_list; do
+		format=$(get_format $v)
+		# if version isn't published or is from a different version, discard
+		if [ -z "$format" ] || [ "$FORMAT" -ne "${format}" ]; then
```
`"$format"` and `"${format}"` are equivalent in this case. It looks like the `"$format"` matches the style of the rest of the file for non-array variable references.

phmccarty commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+		fi
+
+		VERSION[$i]=$v
+		i=$(($i+1))
```
Other test files omit the leading `$` for variable references in arithmetic expressions. For consistency, you can write this as `i=$((i+1))`.

The same comment applies to several other places in this source file, so I won't comment on those lines.

phmccarty commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+# Fails otherwise.
+check_version() {
+	version="$1"
+	cur_version=$(grep VERSION_ID ${ROOT_DIR}/usr/lib/os-release|cut -d = -f 2)
```
optional: I like using AWK instead:

```shell
cur_version=$(awk -F= '$1 == "VERSION_ID" { print $2 }' ${ROOT_DIR}/usr/lib/os-release)
```

otaviobp replied:
For this case I see no reason to use awk, but I know shell-script devs usually hate grep|cut

phmccarty replied:
I'm okay with this change, I just prefer the other way. One improvement would be to grep for `^VERSION_ID=` to avoid matching some other key (or value).

jwakre commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+	echo "Install all packages"
+	all_packages=$(sudo $SWUPD bundle-list --all $SWUPD_OPTS)
+	all_packages=$(echo $all_packages) #take \n out
+	run sudo sh -c "$SWUPD bundle-add $SWUPD_OPTS $all_packages"
```
Usually after a 'run' command, the following line calls an assert (assert_status_is 0).

jwakre commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+	while [ ! -z ${VERSION[$i]} ]; do
+		version="${VERSION[$i]}"
+		i=$(($i+1))
+	done
```
To get the last item in the array, you could do one of the following instead of using a loop:

${VERSION[-1]} # requires bash > v4.1
${VERSION[${#VERSION[@]} -1]}

https://unix.stackexchange.com/questions/198787/is-there-a-way-of-reading-the-last-element-of-an-array-with-bash

otaviobp replied:
hahaha. I should have googled that.

jwakre commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+	done
+}
+
+test_setup() {
```
I think some of the contents of test_setup only need to be run once and could be moved to the global_setup function.

otaviobp replied:
I agree, but I couldn't find an easy way to do that.
Is there a way to export an env variable from one test to the other?

jwakre replied:
I see what you mean. The env variables don't persist between tests... Currently, I think the only ways to export variables between the tests are to call the set_env_variables function in setup() or to give them global scope. Maybe we could add a function to the testlib in the future to make this easier

jwakre replied:
Now, I think this is fine as is. It would be useful to add a feature to testlib to solve this issue in the future.

jwakre commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+}
+
+test_teardown() {
+	if [ "$DEBUG_TEST" != true ]; then
```
I don't think this is very useful because it will only keep the state of the last test to run (update from first to last). For instance, if one of the 'incremental updates' tests fail, its test directory will be deleted by the 'update from first to last' test setup.

otaviobp replied:
I followed the same approach of the teardown in the testlib. So we have a problem in all tests with more than one test when DEBUG_TEST is used.

otaviobp commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,211 @@
+# get_format_from_versions: Echos the first format available in a version from
+# a list of versions
+get_format_from_versions() {
+	for v in "$@"; do
```
I really didn't know that. Thanks

otaviobp commented:
Fixed proposed changes, added --no-scripts to verify and update commands, added a new variable $BUNDLE_LIST to make it possible to run a faster test and added a new test to check if fullfile downloads are working.

To run it in the fast mode just do:

    $ BUNDLE_LIST=os-core bats -t test/real_content/test_swupd.bats

castulo commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,234 @@
+	return 0
+}
+
+# get_format_from_versions: Echos the first format available in a version from
```
one of the things I see you are often overlooking is using variables without enclosing them in double quotes.
> Quoting variables prevents word splitting and glob expansion, and prevents the script from breaking when input contains spaces, line feeds, glob characters and such. - https://github.com/koalaman/shellcheck/wiki/SC2086

>

castulo commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,234 @@
+# get_format_from_versions: Echos the first format available in a version from
+# a list of versions
+get_format_from_versions() {
+	for v; do
```
where are you defining variable "v" here?

phmccarty replied:
`for v; do ... done` is equivalent to `for v in "$@"; do ... done`. @otaviobp changed this in response to my review from a couple days ago.

castulo commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,234 @@
+			continue
+		fi
+
+		VERSION[$i]=$v
```
in line 22 you defined a local variable called VERSIONS:
`local VERSIONS # Array of last available versions. First item is the oldest`

but here you are using VERSION and I don't see VERSIONS anywhere else. did you mean to use VERSIONS here?

castulo commented on test/real_content/test_swupd.bats:
```diff
@@ -0,0 +1,234 @@
+# $BUNDLE_LIST: List of bundles to be installed on the system besides os-core
+#               and os-core-update. If empty, install all bundles.
+
+local SWUPD_OPTS # Swupd options to use in tests
```
you are using local to define some variables here, but local is only valid within functions

otaviobp commented:
> Actually one thing I used to do in the last project I worked on, is that we had shellcheck running in our CI server on every bash script and it would reject the patch if it the scripts didn't comply with shellcheck. It would save reviewers time if we do something like that.

I forgot about that again. Can you set that up for travis? Unless it takes a really long time it would be nice to have this run on every script we change.

castulo commented:
@otaviobp there is still one comment you have not addressed though:

> in line 22 you defined a local variable called VERSIONS:
> `local VERSIONS # Array of last available versions. First item is the oldest`
> 
> but here you are using VERSION and I don't see VERSIONS anywhere else. did you mean to use VERSIONS here?

So didn't you mean to call the variable at line 23 as VERSION instead of VERSIONS?

otaviobp commented:
> So didn't you mean to call the variable at line 23 as VERSION instead of VERSIONS?

yes. fixed.

castulo commented:
@otaviobp did you say you ran shellcheck in your script and fixed all the issues reported by it? I just did and I see 24 issues reported by shellcheck. Starting with the variables you are declaring as local at the beginning of the script. _local_ only make sense inside a function.

otaviobp commented:
Ouch. I renamed the file to a name that was in .gitignore. So I sent the outdated version. Thank you for noticing.
I have just pushed a new version with a change to .gitignore.
I'm also always using shellchecker binary with the seds in Makefile because shellcheck can't recognize load and @test.

castulo commented on test/real_content/swupd_update.bats:
```diff
@@ -0,0 +1,235 @@
+export SWUPD_OPTS # Swupd options to use in tests
+export SWUPD_OPTS_SHORT # Swupd options to use in tests (for bundle-list)
+export FORMAT # Format of the last available version
+export VERSIONS # Array of last available versions. First item is the oldest
```
lol, now VERSIONS is back instead of VERSION 😂

otaviobp replied:
hahahaha

otaviobp commented:
well, I think we're fine now
